File: modules/spotify-remote/ios/SpotifyRemoteModule.swift
import ExpoModulesCore

// Keys for storing access token
private let kSpotifyAccessToken = "spotify_remote_access_token"

/**
 * Shared Spotify manager singleton for iOS
 * Manages SPTAppRemote connection and delegates
 */
class SpotifyManager: NSObject {
    static let shared = SpotifyManager()
    
    var appRemote: SPTAppRemote?
    var configuration: SPTConfiguration?
    var accessToken: String?
    var pendingConnectionPromise: Promise?
    var playerStateSubscription: Bool = false
    var playURI: String = ""
    
    // Event emitter reference
    weak var remoteModule: SpotifyRemoteModule?
    
    // UserDefaults for token persistence
    private var defaults: UserDefaults {
        return UserDefaults.standard
    }
    
    private override init() {
        super.init()
        // Load stored access token
        accessToken = defaults.string(forKey: kSpotifyAccessToken)
    }
    
    func configure(clientId: String, redirectUri: String) {
        guard let url = URL(string: redirectUri) else {
            print("[SpotifyRemote] Invalid redirect URI: \(redirectUri)")
            return
        }
        
        configuration = SPTConfiguration(clientID: clientId, redirectURL: url)
        
        appRemote = SPTAppRemote(configuration: configuration!, logLevel: .debug)
        appRemote?.delegate = self
        
        // Set stored access token if available
        if let token = accessToken {
            appRemote?.connectionParameters.accessToken = token
        }
        
        print("[SpotifyRemote] Configured with clientId: \(clientId)")
    }
    
    func saveAccessToken(_ token: String) {
        accessToken = token
        defaults.set(token, forKey: kSpotifyAccessToken)
        defaults.synchronize()
        appRemote?.connectionParameters.accessToken = token
        print("[SpotifyRemote] Access token saved")
    }
    
    func clearAccessToken() {
        accessToken = nil
        defaults.removeObject(forKey: kSpotifyAccessToken)
        defaults.synchronize()
        print("[SpotifyRemote] Access token cleared")
    }
    
    var isConnected: Bool {
        return appRemote?.isConnected ?? false
    }
    
    /// Handle URL callback from Spotify app
    func handleOpenURL(_ url: URL) -> Bool {
        guard let appRemote = appRemote else { return false }
        
        let parameters = appRemote.authorizationParameters(from: url)
        
        if let accessToken = parameters?[SPTAppRemoteAccessTokenKey] {
            saveAccessToken(accessToken)
            appRemote.connect()
            return true
        } else if let errorDescription = parameters?[SPTAppRemoteErrorDescriptionKey] {
            print("[SpotifyRemote] Auth error: \(errorDescription)")
            pendingConnectionPromise?.reject("AUTH_ERROR", errorDescription)
            pendingConnectionPromise = nil
            return true
        }
        
        return false
    }
}

// MARK: - SPTAppRemoteDelegate
extension SpotifyManager: SPTAppRemoteDelegate {
    func appRemoteDidEstablishConnection(_ appRemote: SPTAppRemote) {
        print("[SpotifyRemote] Connected to Spotify!")
        
        remoteModule?.sendEvent("onConnectionStatusChanged", [
            "connected": true,
            "status": "connected"
        ])
        
        pendingConnectionPromise?.resolve(["connected": true])
        pendingConnectionPromise = nil
    }
    
    func appRemote(_ appRemote: SPTAppRemote, didFailConnectionAttemptWithError error: Error?) {
        print("[SpotifyRemote] Connection failed: \(error?.localizedDescription ?? "Unknown error")")
        
        remoteModule?.sendEvent("onConnectionStatusChanged", [
            "connected": false,
            "status": "failed",
            "error": error?.localizedDescription ?? "Unknown error"
        ])
        
        pendingConnectionPromise?.reject("CONNECTION_FAILED", error?.localizedDescription ?? "Connection failed")
        pendingConnectionPromise = nil
    }
    
    func appRemote(_ appRemote: SPTAppRemote, didDisconnectWithError error: Error?) {
        print("[SpotifyRemote] Disconnected: \(error?.localizedDescription ?? "No error")")
        
        remoteModule?.sendEvent("onConnectionStatusChanged", [
            "connected": false,
            "status": "disconnected",
            "error": error?.localizedDescription as Any
        ])
    }
}

// MARK: - SPTAppRemotePlayerStateDelegate
extension SpotifyManager: SPTAppRemotePlayerStateDelegate {
    func playerStateDidChange(_ playerState: SPTAppRemotePlayerState) {
        guard playerStateSubscription else { return }
        
        let track = playerState.track
        let stateDict: [String: Any] = [
            "track": [
                "uri": track.uri,
                "name": track.name,
                "artist": track.artist.name,
                "album": track.album.name,
                "duration": track.duration,
                "imageUri": track.imageIdentifier ?? ""
            ],
            "playbackPosition": playerState.playbackPosition,
            "playbackSpeed": playerState.playbackSpeed,
            "isPaused": playerState.isPaused,
            "playbackOptions": [
                "isShuffling": playerState.playbackOptions.isShuffling,
                "repeatMode": playerState.playbackOptions.repeatMode.rawValue
            ],
            "playbackRestrictions": [
                "canSkipNext": playerState.playbackRestrictions.canSkipNext,
                "canSkipPrevious": playerState.playbackRestrictions.canSkipPrevious,
                "canSeek": playerState.playbackRestrictions.canSeek,
                "canToggleShuffle": playerState.playbackRestrictions.canToggleShuffle,
                "canRepeatTrack": playerState.playbackRestrictions.canRepeatTrack,
                "canRepeatContext": playerState.playbackRestrictions.canRepeatContext
            ]
        ]
        
        remoteModule?.sendEvent("onPlayerStateChanged", stateDict)
    }
}

// MARK: - SpotifyRemoteModule
public class SpotifyRemoteModule: Module {
    private var manager: SpotifyManager {
        return SpotifyManager.shared
    }
    
    public func definition() -> ModuleDefinition {
        Name("SpotifyRemote")
        
        Events("onPlayerStateChanged", "onConnectionStatusChanged", "onError")
        
        OnCreate {
            SpotifyManager.shared.remoteModule = self
        }
        
        Function("configure") { (clientId: String, redirectUri: String) in
            self.manager.configure(clientId: clientId, redirectUri: redirectUri)
        }
        
        AsyncFunction("connect") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote else {
                promise.reject("NOT_CONFIGURED", "SpotifyRemote not configured. Call configure() first.")
                return
            }
            
            if appRemote.isConnected {
                promise.resolve(["connected": true])
                return
            }
            
            self.manager.pendingConnectionPromise = promise
            
            // Check if Spotify is installed
            guard let spotifyURL = URL(string: "spotify:"),
                  UIApplication.shared.canOpenURL(spotifyURL) else {
                promise.reject("SPOTIFY_NOT_INSTALLED", "Spotify app is not installed on this device")
                self.manager.pendingConnectionPromise = nil
                return
            }
            
            // If we have an access token, try to connect directly
            if self.manager.accessToken != nil {
                appRemote.connect()
            } else {
                // Need to authorize first - this opens Spotify app
                appRemote.authorizeAndPlayURI(self.manager.playURI)
            }
        }
        
        AsyncFunction("disconnect") { (promise: Promise) in
            if let appRemote = self.manager.appRemote, appRemote.isConnected {
                appRemote.disconnect()
            }
            
            self.sendEvent("onConnectionStatusChanged", [
                "connected": false,
                "status": "disconnected"
            ])
            
            promise.resolve(["disconnected": true])
        }
        
        Function("isConnected") { () -> Bool in
            return self.manager.isConnected
        }
        
        AsyncFunction("resume") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.resume { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("pause") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.pause { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("skipToNext") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.skip(toNext: { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            })
        }
        
        AsyncFunction("skipToPrevious") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.skip(toPrevious: { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            })
        }
        
        AsyncFunction("seekTo") { (positionMs: Int, promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.seek(toPosition: positionMs) { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("play") { (uri: String, promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.play(uri) { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("enqueue") { (uri: String, promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.enqueueTrackUri(uri) { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("getPlayerState") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.getPlayerState { playerState, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                    return
                }
                
                guard let state = playerState else {
                    promise.reject("PLAYBACK_ERROR", "Failed to get player state")
                    return
                }
                
                let track = state.track
                let stateDict: [String: Any] = [
                    "track": [
                        "uri": track.uri,
                        "name": track.name,
                        "artist": track.artist.name,
                        "album": track.album.name,
                        "duration": track.duration,
                        "imageUri": track.imageIdentifier ?? ""
                    ],
                    "playbackPosition": state.playbackPosition,
                    "playbackSpeed": state.playbackSpeed,
                    "isPaused": state.isPaused,
                    "playbackOptions": [
                        "isShuffling": state.playbackOptions.isShuffling,
                        "repeatMode": state.playbackOptions.repeatMode.rawValue
                    ],
                    "playbackRestrictions": [
                        "canSkipNext": state.playbackRestrictions.canSkipNext,
                        "canSkipPrevious": state.playbackRestrictions.canSkipPrevious,
                        "canSeek": state.playbackRestrictions.canSeek,
                        "canToggleShuffle": state.playbackRestrictions.canToggleShuffle,
                        "canRepeatTrack": state.playbackRestrictions.canRepeatTrack,
                        "canRepeatContext": state.playbackRestrictions.canRepeatContext
                    ]
                ]
                
                promise.resolve(stateDict)
            }
        }
        
        AsyncFunction("subscribeToPlayerState") { (promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.delegate = self.manager
            appRemote.playerAPI?.subscribe(toPlayerState: { _, error in
                if let error = error {
                    promise.reject("SUBSCRIPTION_ERROR", error.localizedDescription)
                } else {
                    SpotifyManager.shared.playerStateSubscription = true
                    promise.resolve(["subscribed": true])
                }
            })
        }
        
        AsyncFunction("unsubscribeFromPlayerState") { (promise: Promise) in
            SpotifyManager.shared.playerStateSubscription = false
            
            if let appRemote = self.manager.appRemote, appRemote.isConnected {
                appRemote.playerAPI?.unsubscribe(toPlayerState: { _, _ in })
            }
            
            promise.resolve(["unsubscribed": true])
        }
        
        AsyncFunction("setShuffle") { (enabled: Bool, promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            appRemote.playerAPI?.setShuffle(enabled) { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
        
        AsyncFunction("setRepeatMode") { (mode: Int, promise: Promise) in
            guard let appRemote = self.manager.appRemote, appRemote.isConnected else {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify")
                return
            }
            
            let repeatMode: SPTAppRemotePlaybackOptionsRepeatMode
            switch mode {
            case 1:
                repeatMode = .context
            case 2:
                repeatMode = .track
            default:
                repeatMode = .off
            }
            
            appRemote.playerAPI?.setRepeatMode(repeatMode) { _, error in
                if let error = error {
                    promise.reject("PLAYBACK_ERROR", error.localizedDescription)
                } else {
                    promise.resolve(["success": true])
                }
            }
        }
    }
}

File: modules/spotify-remote/android/src/main/java/expo/modules/spotifyremote/SpotifyRemoteModule.kt
package expo.modules.spotifyremote

import android.content.Context
import android.graphics.Bitmap
import android.util.Base64
import android.util.Log
import com.spotify.android.appremote.api.ConnectionParams
import com.spotify.android.appremote.api.Connector
import com.spotify.android.appremote.api.SpotifyAppRemote
import com.spotify.protocol.client.Subscription
import com.spotify.protocol.types.ListItem
import com.spotify.protocol.types.ListItems
import com.spotify.protocol.types.PlayerState
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.ByteArrayOutputStream

/**
 * Expo module for Spotify Remote control on Android
 * Provides playback control, player state observation, and connection management
 */
class SpotifyRemoteModule : Module() {
    companion object {
        private const val TAG = "SpotifyRemote"
    }

    // Spotify App Remote instance
    private var spotifyAppRemote: SpotifyAppRemote? = null
    private var isSubscribed: Boolean = false

    // Configuration
    private var clientId: String = ""
    private var redirectUri: String = ""

    // Pending connection promise
    private var pendingConnectionPromise: Promise? = null

    private val context: Context
        get() = requireNotNull(appContext.reactContext)

    override fun definition() = ModuleDefinition {
        Name("SpotifyRemote")

        // Events that can be sent to JavaScript
        Events("onPlayerStateChanged", "onConnectionStatusChanged", "onError")

        // Configure the module with Spotify credentials
        Function("configure") { clientId: String, redirectUri: String ->
            this@SpotifyRemoteModule.clientId = clientId
            this@SpotifyRemoteModule.redirectUri = redirectUri
        }

        // Connect to Spotify app
        AsyncFunction("connect") { promise: Promise ->
            if (clientId.isEmpty() || redirectUri.isEmpty()) {
                promise.reject("NOT_CONFIGURED", "SpotifyRemote not configured. Call configure() first.", null)
                return@AsyncFunction
            }

            spotifyAppRemote?.let {
                if (it.isConnected) {
                    promise.resolve(mapOf("connected" to true))
                    return@AsyncFunction
                }
            }

            pendingConnectionPromise = promise

            // Check if Spotify is installed before attempting connection
            if (!SpotifyAppRemote.isSpotifyInstalled(context)) {
                Log.e(TAG, "Spotify app is not installed")
                promise.reject("SPOTIFY_NOT_INSTALLED", "Spotify app is not installed on this device", null)
                pendingConnectionPromise = null
                return@AsyncFunction
            }

            val connectionParams = ConnectionParams.Builder(clientId)
                .setRedirectUri(redirectUri)
                .showAuthView(true)
                .build()

            Log.d(TAG, "Connecting to Spotify with clientId: $clientId, redirectUri: $redirectUri")

            SpotifyAppRemote.connect(context, connectionParams, object : Connector.ConnectionListener {
                override fun onConnected(appRemote: SpotifyAppRemote) {
                    spotifyAppRemote = appRemote
                    Log.d(TAG, "Connected to Spotify!")

                    sendEvent("onConnectionStatusChanged", mapOf(
                        "connected" to true,
                        "status" to "connected"
                    ))

                    pendingConnectionPromise?.resolve(mapOf("connected" to true))
                    pendingConnectionPromise = null
                }

                override fun onFailure(throwable: Throwable) {
                    Log.e(TAG, "Connection failed: ${throwable.message}", throwable)

                    sendEvent("onConnectionStatusChanged", mapOf(
                        "connected" to false,
                        "status" to "failed",
                        "error" to (throwable.message ?: "Unknown error")
                    ))

                    pendingConnectionPromise?.reject("CONNECTION_FAILED", throwable.message, throwable)
                    pendingConnectionPromise = null
                }
            })
        }

        // Connect silently to Spotify app (without opening Spotify)
        // On Android, this attempts to connect to an already-running Spotify instance
        AsyncFunction("connectSilent") { promise: Promise ->
            if (clientId.isEmpty() || redirectUri.isEmpty()) {
                promise.reject("NOT_CONFIGURED", "SpotifyRemote not configured. Call configure() first.", null)
                return@AsyncFunction
            }

            spotifyAppRemote?.let {
                if (it.isConnected) {
                    promise.resolve(mapOf("connected" to true))
                    return@AsyncFunction
                }
            }

            // Check if Spotify is installed
            if (!SpotifyAppRemote.isSpotifyInstalled(context)) {
                Log.e(TAG, "Spotify app is not installed")
                promise.reject("SPOTIFY_NOT_INSTALLED", "Spotify app is not installed on this device", null)
                return@AsyncFunction
            }

            // Try to connect silently without showing auth view
            val connectionParams = ConnectionParams.Builder(clientId)
                .setRedirectUri(redirectUri)
                .showAuthView(false)
                .build()

            Log.d(TAG, "Attempting silent connection to Spotify...")

            SpotifyAppRemote.connect(context, connectionParams, object : Connector.ConnectionListener {
                override fun onConnected(appRemote: SpotifyAppRemote) {
                    spotifyAppRemote = appRemote
                    Log.d(TAG, "Silent connection to Spotify successful!")

                    sendEvent("onConnectionStatusChanged", mapOf(
                        "connected" to true,
                        "status" to "connected"
                    ))

                    promise.resolve(mapOf("connected" to true))
                }

                override fun onFailure(throwable: Throwable) {
                    Log.d(TAG, "Silent connection failed: ${throwable.message}")
                    // Silent connection failed - this is expected if Spotify isn't running
                    // Don't emit error event for silent failures
                    promise.resolve(mapOf("connected" to false))
                }
            })
        }

        // Disconnect from Spotify app
        AsyncFunction("disconnect") { promise: Promise ->
            spotifyAppRemote?.let {
                SpotifyAppRemote.disconnect(it)
                spotifyAppRemote = null
            }

            sendEvent("onConnectionStatusChanged", mapOf(
                "connected" to false,
                "status" to "disconnected"
            ))

            promise.resolve(mapOf("disconnected" to true))
        }

        // Check if connected
        Function("isConnected") {
            spotifyAppRemote?.isConnected ?: false
        }

        // Resume playback
        AsyncFunction("resume") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.resume()
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Pause playback
        AsyncFunction("pause") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.pause()
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Skip to next track
        AsyncFunction("skipToNext") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.skipNext()
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Skip to previous track
        AsyncFunction("skipToPrevious") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.skipPrevious()
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Seek to position (in milliseconds)
        AsyncFunction("seekTo") { positionMs: Long, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.seekTo(positionMs)
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Play a specific track, album, or playlist by URI
        AsyncFunction("play") { uri: String, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.play(uri)
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Enqueue a track (add to queue)
        AsyncFunction("enqueue") { uri: String, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.queue(uri)
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Get current player state
        AsyncFunction("getPlayerState") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.playerState
                .setResultCallback { playerState ->
                    promise.resolve(playerStateToMap(playerState))
                }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Subscribe to player state changes
        AsyncFunction("subscribeToPlayerState") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            if (isSubscribed) {
                // Already subscribed
                promise.resolve(mapOf("subscribed" to true))
                return@AsyncFunction
            }

            remote.playerApi.subscribeToPlayerState()
                .setEventCallback { playerState ->
                    sendEvent("onPlayerStateChanged", playerStateToMap(playerState))
                }
                .setLifecycleCallback(object : Subscription.LifecycleCallback {
                    override fun onStart() {
                        isSubscribed = true
                        Log.d(TAG, "Player state subscription started")
                    }
                    override fun onStop() {
                        isSubscribed = false
                        Log.d(TAG, "Player state subscription stopped")
                    }
                })
                .setErrorCallback { throwable ->
                    sendEvent("onError", mapOf("error" to (throwable.message ?: "Unknown error")))
                }

            isSubscribed = true
            promise.resolve(mapOf("subscribed" to true))
        }

        // Unsubscribe from player state changes
        AsyncFunction("unsubscribeFromPlayerState") { promise: Promise ->
            isSubscribed = false
            // Note: The subscription is managed internally by SpotifyAppRemote
            // and will be cancelled when we disconnect
            promise.resolve(mapOf("unsubscribed" to true))
        }

        // Set shuffle mode
        AsyncFunction("setShuffle") { enabled: Boolean, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.setShuffle(enabled)
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // Set repeat mode: 0 = off, 1 = context, 2 = track
        AsyncFunction("setRepeatMode") { mode: Int, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.playerApi.setRepeat(mode)
                .setResultCallback { promise.resolve(mapOf("success" to true)) }
                .setErrorCallback { throwable ->
                    promise.reject("PLAYBACK_ERROR", throwable.message, throwable)
                }
        }

        // ==================== Content API ====================

        // Get recommended content items (playlists, albums, etc.)
        AsyncFunction("getRecommendedContentItems") { promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            remote.contentApi.getRecommendedContentItems("default")
                .setResultCallback { listItems: ListItems ->
                    val items = listItems.items.map { listItemToMap(it) }
                    promise.resolve(items)
                }
                .setErrorCallback { throwable ->
                    Log.e(TAG, "Failed to get recommended content: ${throwable.message}", throwable)
                    promise.reject("CONTENT_ERROR", throwable.message, throwable)
                }
        }

        // Get children of a content item (e.g., tracks in a playlist)
        AsyncFunction("getChildrenOfContentItem") { uri: String, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            // Create a ListItem to get children from
            remote.contentApi.getChildrenOfItem(ListItem(uri, uri, null, null, null, false, true), 50, 0)
                .setResultCallback { listItems: ListItems ->
                    val items = listItems.items.map { listItemToMap(it) }
                    promise.resolve(items)
                }
                .setErrorCallback { throwable ->
                    Log.e(TAG, "Failed to get children of content: ${throwable.message}", throwable)
                    promise.reject("CONTENT_ERROR", throwable.message, throwable)
                }
        }

        // Get image for a content item as base64 (pass the imageUri from the content item)
        AsyncFunction("getContentItemImage") { imageUri: String, promise: Promise ->
            val remote = spotifyAppRemote
            if (remote == null || !remote.isConnected) {
                promise.reject("NOT_CONNECTED", "Not connected to Spotify", null)
                return@AsyncFunction
            }

            // On Android, we receive the imageUri directly from the ListItem
            // Create an ImageUri and fetch the image
            val spotifyImageUri = com.spotify.protocol.types.ImageUri(imageUri)
            
            remote.imagesApi.getImage(spotifyImageUri)
                .setResultCallback { bitmap: Bitmap ->
                    // Convert bitmap to base64
                    val outputStream = ByteArrayOutputStream()
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 80, outputStream)
                    val base64String = Base64.encodeToString(outputStream.toByteArray(), Base64.NO_WRAP)
                    promise.resolve("data:image/jpeg;base64,$base64String")
                }
                .setErrorCallback { throwable ->
                    Log.e(TAG, "Failed to get content image: ${throwable.message}", throwable)
                    promise.reject("IMAGE_ERROR", throwable.message, throwable)
                }
        }

        // Handle app lifecycle
        OnActivityEntersForeground {
            // Attempt reconnection when app comes to foreground
            spotifyAppRemote?.let {
                if (!it.isConnected && clientId.isNotEmpty()) {
                    // Could auto-reconnect here if needed
                }
            }
        }

        OnDestroy {
            isSubscribed = false
            spotifyAppRemote?.let { SpotifyAppRemote.disconnect(it) }
        }
    }

    // Convert PlayerState to Map for JavaScript
    private fun playerStateToMap(playerState: PlayerState): Map<String, Any?> {
        val track = playerState.track
        return mapOf(
            "track" to if (track != null) mapOf(
                "uri" to track.uri,
                "name" to track.name,
                "artist" to track.artist?.name,
                "album" to track.album?.name,
                "duration" to track.duration,
                "imageUri" to track.imageUri?.raw
            ) else null,
            "playbackPosition" to playerState.playbackPosition,
            "playbackSpeed" to playerState.playbackSpeed,
            "isPaused" to playerState.isPaused,
            "playbackOptions" to mapOf(
                "isShuffling" to playerState.playbackOptions.isShuffling,
                "repeatMode" to playerState.playbackOptions.repeatMode
            ),
            "playbackRestrictions" to mapOf(
                "canSkipNext" to playerState.playbackRestrictions.canSkipNext,
                "canSkipPrevious" to playerState.playbackRestrictions.canSkipPrev,
                "canSeek" to playerState.playbackRestrictions.canSeek,
                "canToggleShuffle" to playerState.playbackRestrictions.canToggleShuffle,
                "canRepeatTrack" to playerState.playbackRestrictions.canRepeatTrack,
                "canRepeatContext" to playerState.playbackRestrictions.canRepeatContext
            )
        )
    }

    // Convert ListItem to Map for JavaScript (Content API)
    private fun listItemToMap(item: ListItem): Map<String, Any?> {
        return mapOf(
            "uri" to item.uri,
            "title" to item.title,
            "subtitle" to item.subtitle,
            "imageUri" to item.imageUri?.raw,
            "isPlayable" to item.playable,
            "isContainer" to item.hasChildren
        )
    }
}
